File: WOT-iOS/WOT-iOS/core/datasources/WOTTree/WOTTree+Pivot/WOTPivotNode.m
//
//  WOTPivotNode.m
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 8/27/15.
//  Copyright (c) 2015 Pavel Yeshchyk. All rights reserved.
//

#import "WOTPivotNode.h"

@implementation WOTPivotNode


#pragma mark -
- (id)copyWithZone:(NSZone *)zone {
    
    WOTPivotNode *result = [super copyWithZone:zone];
    
    result.dataColor = self.dataColor;
    result.predicate = self.predicate;
    return result;
}

- (id)copyWithPredicate:(NSPredicate *)predicate {

    WOTPivotNode *result = [self copy];
    result.predicate = predicate;
    return result;
}

- (id)initWithName:(NSString *)name predicate:(NSPredicate *)predicate {
    
    self = [self initWithName:name];
    if (self) {
        
        [self setPredicate:predicate];
    }
    return self;
}

- (id)initWithName:(NSString *)name imageURL:(NSURL *)imageURL predicate:(NSPredicate *)predicate {
    
    self = [self initWithName:name imageURL:imageURL];
    if (self){
        
        [self setPredicate:predicate];
    }
    return self;
}

- (id)initWithName:(NSString *)name isVisible:(BOOL)isVisible {
    
    self = [self initWithName:name];
    if (self){
        
        self.isVisible = isVisible;
    }
    return self;
}

- (void)dealloc {
    
}

#pragma mark - getters / setters

- (PivotStickyType)stickyType {
    
    return PivotStickyTypeFloat;
}

@end

File: WOT-iOS/WOT-iOS/core/datasources/WOTTree/WOTTree+Pivot/WOTPivotNode.swift
//
//  WOTPivotNode.swift
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 7/17/18.
//  Copyright © 2018 Pavel Yeshchyk. All rights reserved.
//

import Foundation
import CoreData

@objc
protocol WOTPivotNodeProtocol: WOTNodeProtocol {
    var dataColor: UIColor { get set }
    var data1: NSManagedObject? { get set }
    var stickyType: PivotStickyType  { get }
    var predicate: NSPredicate?  { get set }
}

class WOTPivotNodeSwift: WOTNodeSwift, WOTPivotNodeProtocol {

    var dataColor: UIColor = UIColor.clear
    var data1: NSManagedObject?
    var stickyType: PivotStickyType { return .float }
    var predicate: NSPredicate?

    @objc
    convenience init(name nameValue: String, predicate predicateValue: NSPredicate) {
        self.init(name: nameValue)
        self.predicate = predicateValue
    }

    public override func copy(with zone: NSZone? = nil) -> Any {
        let result = type(of: self).init(name: self.name)
        result.predicate = self.predicate
        result.dataColor = self.dataColor
        result.isVisible = self.isVisible
        result.imageURL = self.imageURL
        return result;
    }
}
